crates/nargo_cli/src/cli/test_cmd.rs

```diff
@@ -123,30 +122,56 @@ fn run_tests<B: Backend>(
 fn run_test<B: Backend>(
     backend: &B,
     test_name: &str,
-    main: FuncId,
+    test_function: TestFunction,
     context: &Context,
     show_output: bool,
     config: &CompileOptions,
 ) -> Result<(), CliError<B>> {
-    let mut program = compile_no_check(context, config, main).map_err(|err| {
+    let report_error = |err| {
         noirc_errors::reporter::report_all(&context.file_manager, &[err], config.deny_warnings);
-        CliError::Generic(format!("Test '{test_name}' failed to compile"))
-    })?;
-
-    // Note: We could perform this test using the unoptimized ACIR as generated by `compile_no_check`.
-    program.circuit = optimize_circuit(backend, program.circuit).unwrap().0;
-
-    // Run the backend to ensure the PWG evaluates functions like std::hash::pedersen,
-    // otherwise constraints involving these expressions will not error.
-    match execute_circuit(backend, program.circuit, WitnessMap::new(), show_output) {
-        Ok(_) => Ok(()),
-        Err(error) => {
-            let writer = StandardStream::stderr(ColorChoice::Always);
-            let mut writer = writer.lock();
-            writer.set_color(ColorSpec::new().set_fg(Some(Color::Red))).ok();
-            writeln!(writer, "failed").ok();
-            writer.reset().ok();
-            Err(error.into())
+        Err(CliError::Generic(format!("Test '{test_name}' failed to compile")))
+    };
+
+    let program = compile_no_check(context, config, test_function.get_id());
+    match program {
+        Ok(mut program) => {
+            // Note: We don't need to use the optimized ACIR here
+            program.circuit = optimize_circuit(backend, program.circuit).unwrap().0;
+
+            // Run the backend to ensure the PWG evaluates functions like std::hash::pedersen,
+            // otherwise constraints involving these expressions will not error.
+            let circuit_execution =
+                execute_circuit(backend, program.circuit, WitnessMap::new(), show_output);
+
+            if test_function.get_expect_failure_flag() {
```

Comment:
```suggestion
            if test_function.should_fail() {
```
We should name this better, as it is very confusing